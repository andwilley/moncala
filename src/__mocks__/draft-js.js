import React from 'react';

// mock the editor for unit tests
export class Editor extends React.Component {
  focus() {return undefined}
  render() {return null}
}

export const EditorState = {
  createEmpty: () => '',
  createWithContent: () => null,
  getCurrentContent: () => ({
    getPlainText: () => ''
  })
}

export const ContentState = {
  createFromText: text => text
}