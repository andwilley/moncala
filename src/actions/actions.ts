import { action, ActionType } from 'typesafe-actions';
import { ThunkAction } from 'redux-thunk';
import { IState } from '../types/IState';
import { MoncalaDb } from '../indexeddb/db';
import { INote } from '../api/models';
import day from 'dayjs';

// Action type constants

export const SET_NOTES = 'SET_NOTES';
export const ADD_NOTE = 'ADD_NOTE';
export const UPDATE_NOTE = 'UPDATE_NOTE';

// Action creators

export const setNotes = (notes: INote[]) => action(SET_NOTES, notes);
export const loadNotes = (): ThunkAction<void, IState, MoncalaDb, Action> => async (
  dispatch,
  _,
  db
) => {
  const notes = await db.notes.toArray();
  dispatch(setNotes(notes));
};

export const addNote = (note: INote) => action(ADD_NOTE, note);
export const addNoteAsync = (
  content: string
): ThunkAction<void, IState, MoncalaDb, Action> => async (dispatch, _, db) => {
  const newNote: INote = {
    dateCreated: day().toISOString(),
    dateModified: day().toISOString(),
    content
  };
  const newId = await db.notes.add(newNote);
  dispatch(addNote({ ...newNote, id: newId }));
};

export const updateNote = (noteId: number, note: Partial<INote>) =>
  action(UPDATE_NOTE, { id: noteId, note });

export type Action =
  | ActionType<typeof addNote>
  | ActionType<typeof updateNote>
  | ActionType<typeof setNotes>;
