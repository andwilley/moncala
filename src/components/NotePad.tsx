import React from "react";

export const NotePad: React.FC = () => {
  return <div className="Notes dib w-100 pa3 bl b--near-white br3 br--left shadow-2">Notes</div>;
};

export default NotePad;
