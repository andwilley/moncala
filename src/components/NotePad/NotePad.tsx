import React, { useRef, useEffect } from 'react';
import { Editor, EditorState } from 'draft-js';
import { useSelector, useDispatch } from 'react-redux';
import { getEditorState, getActiveId, getFocusEditor } from '../../reducers/reducer';
import {
  setEditorState,
  addNoteAsync,
  updateNoteAsync,
  setFocusEditor
} from '../../actions/actions';

export const NotePad: React.FC = () => {
  const editorState = useSelector(getEditorState);
  const activeId = useSelector(getActiveId);
  const focusEditor = useSelector(getFocusEditor);
  const dispatch = useDispatch();
  const editorRef = useRef<Editor>(null);

  // syncronize this component with the active id
  useEffect(() => {
    if (focusEditor && editorRef && editorRef.current) {
      editorRef.current.focus();
      dispatch(setFocusEditor(false));
    }
  }, [activeId, focusEditor, dispatch]);

  /**
   * On click in the surrounding div, focus the editor
   * If no note is active, add a new note
   */
  const onEditAreaClick = (e: React.MouseEvent) => {
    if (editorRef && editorRef.current) {
      editorRef.current.focus();
      if (!activeId) {
        dispatch(addNoteAsync());
      }
    }
  };

  /**
   * manage controlled component editor
   * @param editState the draft editor state
   */
  const onEditorChange = (editState: EditorState) => {
    dispatch(setEditorState(editState));
    const text = editState.getCurrentContent().getPlainText();
    if (activeId) {
      dispatch(updateNoteAsync(activeId, text));
    }
  };

  return (
    <div
      title="NotePad"
      className="NotePad dib w-100 pa3 bl b--near-white br3 br--left shadow-1"
      onClick={onEditAreaClick}
    >
      <Editor editorState={editorState} onChange={onEditorChange} ref={editorRef} />
    </div>
  );
};

export default NotePad;
