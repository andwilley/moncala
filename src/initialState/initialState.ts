import { IState } from '../types/IState';
import { EditorState } from 'draft-js';

export const initialState: IState = {
  notes: {
    byId: {},
    allIds: []
  },
  activeId: undefined,
  editorState: EditorState.createEmpty(),
  focusEditor: false
};
