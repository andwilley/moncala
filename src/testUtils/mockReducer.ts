import { IState } from '../types/IState';
import { Action } from '../actions/actions';

const mockReducer = (testState: IState) => (
  state: IState | undefined = testState,
  action: Action
) => state;

export default mockReducer;
