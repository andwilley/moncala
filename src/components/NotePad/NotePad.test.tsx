import React from 'react';
import renderer from 'react-test-renderer';
import NotePad from './NotePad';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import mockReducer from '../../testUtils/mockReducer';
import { initialState } from '../../initialState/initialState';
import { Editor, EditorState } from 'draft-js';
import { setEditorState, addNoteAsync } from '../../actions/actions';

describe('<NotePad> test a render output', () => {
  it('should render the notepad', () => {
    const store = createStore(mockReducer(initialState));
    store.dispatch = jest.fn();
    const tree = renderer
      .create(
        <Provider store={store}>
          <NotePad />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call dispatch on editorChange', () => {
    const store = createStore(mockReducer(initialState));
    store.dispatch = jest.fn();

    const comp = renderer.create(
      <Provider store={store}>
        <NotePad />
      </Provider>
    );
    const inst = comp.root;
    const emptyState = EditorState.createEmpty();
    const element = inst.findByType(Editor);
    element.props.onChange(emptyState);
    expect(store.dispatch).toHaveBeenCalledWith(setEditorState(emptyState));
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });

  it('should call dispatch on editorChange', () => {
    const store = createStore(mockReducer({ ...initialState, focusEditor: true, activeId: 1 }));
    store.dispatch = jest.fn();

    const comp = renderer.create(
      <Provider store={store}>
        <NotePad />
      </Provider>
    );
    const inst = comp.root;
    inst.findByType(Editor).props.onChange(EditorState.createEmpty());
    expect(store.dispatch).toHaveBeenCalledTimes(2);
  });

  it('should call dispatch on edit area click', () => {
    const store = createStore(mockReducer(initialState));
    store.dispatch = jest.fn();

    const comp = renderer.create(
      <Provider store={store}>
        <NotePad />
      </Provider>
    );
    const inst = comp.root;
    inst.findByProps({ title: 'NotePad' }).props.onClick({});
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });

  it('should not call dispatch on edit area click if activeId is set', () => {
    const store = createStore(mockReducer({ ...initialState, activeId: 1 }));
    store.dispatch = jest.fn();

    const comp = renderer.create(
      <Provider store={store}>
        <NotePad />
      </Provider>
    );
    const inst = comp.root;
    inst.findByProps({ title: 'NotePad' }).props.onClick({});
    expect(store.dispatch).toHaveBeenCalledTimes(0);
  });
});
