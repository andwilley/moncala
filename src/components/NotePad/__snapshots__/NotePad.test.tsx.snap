// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotePad> test a render output should render the notepad 1`] = `
<div
  className="NotePad dib w-100 pa3 bl b--near-white br3 br--left shadow-1"
  onClick={[Function]}
  title="NotePad"
/>
`;
